stan/math/prim/mat/prob/dirichlet_lpmf.hpp

```diff
@@ -40,41 +51,55 @@ namespace math {
  * @throw std::domain_error if any element of theta is less than 0.
  * @throw std::domain_error if the sum of theta is not 1.
  * @tparam T_prob Type of scalar.
- * @tparam T_prior_sample_size Type of prior sample sizes.
+ * @tparam T_prior_size Type of prior sample sizes.
  */
-template <bool propto, typename T_prob, typename T_prior_sample_size>
-typename boost::math::tools::promote_args<T_prob, T_prior_sample_size>::type
-dirichlet_lpmf(
-    const Eigen::Matrix<T_prob, Eigen::Dynamic, 1>& theta,
-    const Eigen::Matrix<T_prior_sample_size, Eigen::Dynamic, 1>& alpha) {
+template <bool propto, typename T_prob, typename T_prior_size>
+typename return_type<T_prob, T_prior_size>::type dirichlet_lpmf(
+    const T_prob& theta, const T_prior_size& alpha) {
   static const char* function = "dirichlet_lpmf";
-  using boost::math::lgamma;
-  using boost::math::tools::promote_args;
 
-  typename promote_args<T_prob, T_prior_sample_size>::type lp(0.0);
+  typedef typename stan::partials_return_type<T_prob, T_prior_size>::type
+      T_partials_return;
+  typedef typename Eigen::Matrix<T_partials_return, -1, 1> T_partials_vec;
+
   check_consistent_sizes(function, "probabilities", theta, "prior sample sizes",
                          alpha);
   check_positive(function, "prior sample sizes", alpha);
   check_simplex(function, "probabilities", theta);
 
-  if (include_summand<propto, T_prior_sample_size>::value) {
-    lp += lgamma(alpha.sum());
-    for (int k = 0; k < alpha.rows(); ++k)
-      lp -= lgamma(alpha[k]);
-  }
-  if (include_summand<propto, T_prob, T_prior_sample_size>::value) {
-    for (int k = 0; k < theta.rows(); ++k)
-      lp += multiply_log(alpha[k] - 1, theta[k]);
-  }
-  return lp;
+  vector_seq_view<T_prob> theta_vec(theta);
+  T_partials_vec theta_dbl = value_of(theta_vec[0]);
+
+  vector_seq_view<T_prior_size> alpha_vec(alpha);
+  T_partials_vec alpha_dbl = value_of(alpha_vec[0]);
+
+  T_partials_return lp(0.0);
+
+  if (include_summand<propto, T_prior_size>::value)
+    lp += lgamma(alpha_dbl.sum()) - lgamma(alpha_dbl).sum();
+
+  if (include_summand<propto, T_prob, T_prior_size>::value)
+    lp += (theta_dbl.array().log() * (alpha_dbl.array() - 1.0)).sum();
```

Comment:
[comment---no action required]
This doesn't have the behavior of `multiply_log` at (0, 0), but I don't think that's going to be an issue here as everything's going to need to stay non-zero.